function to automatically determine the electricity price based on the country

mahendrark opened the pull request.

SabineHaas commented:
@mahendrark  concerning your question: the first problem you are facing when reading the csv file is that the delimiter is "," as well as the decimal is ",".  I'd recommend you to change the delimiter to ";" and/or the decimal to ".". However, you can also keep the decimal when setting `decimal=","` in `read_csv()`.

Concerning the first steps of processing the data you can try skipping the first rows and drop the all nan columns.
`pd.read_csv(prices_file_path, skiprows=3, index_col=0).dropna(how="all", axis=1)` 

I hope this helps :)

mahendrark commented:
hi @SabineHaas 
thanks for your helpful pointers. 

I have however, modified the csv and deleted the irrelevant rows and data (such as energy prices for all years other than 2019 as that is redundant). Now it works fine.

SabineHaas commented on pvcompare/check_inputs.py:
```diff
@@ -345,9 +400,8 @@ def create_mvs_energy_production_file(pv_setup, energy_production_filename):
 
 def add_parameters_to_energy_production_file(
-    pp_number, ts_filename, nominal_value, mvs_input_directory=None
+        pp_number, ts_filename, nominal_value, mvs_input_directory=None
```
It looks like some spaces occur that shouldn't be there. You might be able to solve that just with black (`black . --exclude docs/`).

SabineHaas commented on pvcompare/check_inputs.py:
```diff
@@ -182,18 +182,74 @@ def add_project_data(mvs_input_directory, latitude, longitude, country, year):
+    This function is called by the main function when then user-input value of the cost of
+    grid electricity is None (i.e., not provided by the user). This function then determines the cost
+    of electricity by checking in the energyProviders.csv, and returns the value.
+    If the value is not provided in that cscv either,
```
```suggestion
    If the value is not provided in that csv either,
```

SabineHaas commented on pvcompare/check_inputs.py:
```diff
@@ -182,18 +182,74 @@ def add_project_data(mvs_input_directory, latitude, longitude, country, year):
+
+    Parameters:
+    -----------
+   mvs_input_directory:
```
Also here the indent does not look right.

SabineHaas commented on pvcompare/check_inputs.py:
```diff
@@ -182,18 +182,74 @@ def add_project_data(mvs_input_directory, latitude, longitude, country, year):
+        mvs_input_directory, "csv_elements/" "energyProviders.csv"
+    )
+
+    # Create dataframe containing the hoursehold electricity prices in the EU nations
```
```suggestion
    # Create dataframe containing the household electricity prices in the EU nations
```

SabineHaas commented on pvcompare/check_inputs.py:
```diff
@@ -182,18 +182,74 @@ def add_project_data(mvs_input_directory, latitude, longitude, country, year):
+
+        if energy_price is None:
+            logging.info(f"The parameter {energy_price} is taken " f"from energyProviders.csv.")
+            energy_price = grid_related.at[energy_price, "Electricity grid"]
```
When running the code locally I get an error here. `KeyError: 'Electricity grid'`
Please check whether you can reproduce and change this.
You can solve this by writing tests for your new code!

mahendrark replied:
@SabineHaas 

I have installation problems with PVCompare. I am unable to run the simulation. I will write an issue soon.

Can you please check it now? It was because of an additional space in the name of the column. I fixed it and pushed now.

SabineHaas replied:
> I have installation problems with PVCompare. I am unable to run the simulation. I will write an issue soon.
> 

Yes, please do that! :)

> Can you please check it now? It was because of an additional space in the name of the column. I fixed it and pushed now.

I still get an error. If you cannot run main.py because of installation issues please test your function separately. For that you don't need all the packages of pvcompare.
You can directly write a test into [`test_check_inputs.py`](https://github.com/greco-project/pvcompare/blob/dev/tests/test_check_inputs.py)

By the way, in the `check_inputs` module there are many unused imports that you can just uncomment with `#` to avoid problems with these packages (no need to install then).

mahendrark replied:
Thank you so much for the suggestion Sabine!
I ran the function separately and tested it with different inputs. 

The problem was with the method calls on one of the dataframes. I used the variable energy_price wrongly. Fixed that and its working as intended now for me. 
I pushed the changes. Let me know if it is alright :)

SabineHaas commented:
Thank you for the updates.

Please write tests for the new function (you can use the tests you did mentioned above). 
Then, before merging all checks need to run through (I think this time it's only black).

mahendrark commented:
@SabineHaas 

Can you guide me with writing the tests?
I cannot figure what to test actually. 

There are three input parameters: `mvs_input_directory, energy_price, country`
Inia already wrote tests for `mvs_input_directory` and `country`. 
So the user can only input right parameters for those two.

As for `energy_price`, if the user inputs a non-numeric value, then MVS would raise an error I think. 
If the user does not input a value via main function, then we check the csv file for a value. If there is none, then we fetch it from the csv with EU electricity prices.

mahendrark commented on pvcompare/check_inputs.py:
```diff
@@ -188,7 +189,75 @@ def add_project_data(mvs_input_directory, latitude, longitude, country, year):
+    Parameters:
+    -----------
+    mvs_input_directory:
+    energy_price: the price of electricity is either None or 0
```
I will remove the 0. Thanks for pointing that out :)

mahendrark commented on pvcompare/check_inputs.py:
```diff
@@ -188,7 +189,75 @@ def add_project_data(mvs_input_directory, latitude, longitude, country, year):
+            logging.info(
+                f"The parameter energy_price is taken from energyProviders.csv."
+            )
+            energy_price = grid_related.at["energy_price", "Electricity grid "]
```
I renamed the parameter `energy_price` as `electricity_price`. Now, we should not have confusion between `electricity_price` (the variable that holds the eöectricity price) and `energy_price` (the index key in energyProviders.csv).

The variable for which the price has been read from "electricity_prices_households.csv" is renamed as `electricity_price_from_csv`. 

Thank you Sabine for pointing this out. :)

mahendrark commented on pvcompare/check_inputs.py:
```diff
@@ -188,7 +189,75 @@ def add_project_data(mvs_input_directory, latitude, longitude, country, year):
+    return latitude, longitude, country, year
+
+
+def energy_price_check(mvs_input_directory, energy_price, country):
```
done :)

mahendrark commented:
@SabineHaas 

I cant test if the KeyError is raised if the user enters a non-EU country because the previous test always overwrites the NaN in the csv with 0.5 value for energy_price. So when this text is executed, it doesnt care for the country and just takes the value of the electricity_price from the energy_price value of 0.5 or whatever we provide in the test function call and overwrites whatever value is in the csv with this value.

mahendrark commented:
@SabineHaas 

the test for checking if the country is valid or not is failing and I cannot find a way to fix it. 
The problem is, when the user doesnt provide a value for the energy price, and provides a wrong country.. 
the check_energy_price doesnt care about these if there is a value in the csv. 

So I tried making the value in the csv as NaN. Then the test passes.

But the other function which checks if the user provided value of electricity_price will be written into the csv energyProviders.csv or not, is overwriting the NaN with the user provided value.

So when you run the test, the second time.. it fails.

SabineHaas commented:
> @SabineHaas
> 
> the test for checking if the country is valid or not is failing and I cannot find a way to fix it.
> The problem is, when the user doesnt provide a value for the energy price, and provides a wrong country..
> the check_energy_price doesnt care about these if there is a value in the csv.
> 
> So I tried making the value in the csv as NaN. Then the test passes.
> 
> But the other function which checks if the user provided value of electricity_price will be written into the csv energyProviders.csv or not, is overwriting the NaN with the user provided value.
> 
> So when you run the test, the second time.. it fails.
> 
> So for the test to pass, you have to set the energy_price value in energyProviders.csv in mvs test data to NaN. Then it works fine.

This calls for a `teardown_method` - you can find an example for example [here](https://github.com/rl-institut/mvs_eland/blob/f402adafaa38255c849eb8636e0a9c97b9b50975/tests/test_B0_data_input_json.py#L86).
You could have a setup_method that copies the csv from somewhere to a test folder, during the test you use it (read, write, etc.) and after the test you can delete it (with teardown_method).

I also saw that you change the country to "Uganda" in the setup of the class. You should only do that if you are sure that this is not needed for the other tests (it was Spain before). If you need a different setup for your tests you should create a new class incl. setup.

mahendrark commented:
@SabineHaas 

Thanks for the info. 
Those are new to me. I will look into them. 

I am closing this merging this branch and closing this PR. I will create a new PR for the tests and use that.

SabineHaas commented:
I will revert this merge as the tests did not pass. 
You can still work in another branch by branching from `read_energy_price` if you prefer to work in a new branch.
